File: react-native/src/push/PushService.ts
/**
 * 推送基线：FCM/APNs 令牌注册 + 前台消息转发。
 * messaging 依赖原生 Firebase 配置（google-services / GoogleService-Info），
 * 与 Telemetry 的 firebase 三件套同策略：动态 import + 失败静默降级，
 * web / 未配置原生工程时推送不可用但不阻塞登录流程。
 */
import { apiClient } from '../data/apiClient';
import { telemetry } from '../telemetry/Telemetry';

// @react-native-firebase/messaging 的最小使用面（与 IAP 懒加载同一模式）。
interface RemoteMessageLike {
  notification?: { title?: string; body?: string };
}
interface MessagingApp {
  requestPermission(): Promise<number>;
  getToken(): Promise<string>;
  deleteToken(): Promise<void>;
  onTokenRefresh(listener: (token: string) => void): () => void;
  onMessage(listener: (message: RemoteMessageLike) => void): () => void;
}
type MessagingFactory = () => MessagingApp;

let messaging: MessagingFactory | null = null;
let loaded = false;
async function loadMessaging(): Promise<MessagingFactory | null> {
  if (loaded) return messaging;
  loaded = true;
  try {
    const module = await import('@react-native-firebase/messaging');
    // 类型层未声明 default 导出（运行时存在），与 IAP 同样走最小使用面断言。
    messaging = (module as unknown as { default?: MessagingFactory }).default ?? null;
  } catch (error) {
    telemetry.report(error instanceof Error ? error : new Error('messaging unavailable'));
  }
  return messaging;
}

let unsubscribe: (() => void) | null = null;
let token: string | null = null;

export async function startPush(onForeground: (message: string) => void): Promise<void> {
  const factory = await loadMessaging();
  if (!factory || unsubscribe) return;
  try {
    await factory().requestPermission();
  } catch {
    return; // 权限被拒：推送降级为不可用，不重复注册
  }
  const instance = factory();
  token = await instance.getToken();
  await apiClient.registerPushToken(token);
  const tokenSub = instance.onTokenRefresh((next) => {
    token = next;
    apiClient.registerPushToken(next).catch((error) => {
      telemetry.report(error instanceof Error ? error : new Error('token refresh failed'));
    });
  });
  const messageSub = instance.onMessage((message) => {
    const body = message.notification?.body;
    if (body) onForeground(body);
  });
  unsubscribe = () => {
    tokenSub();
    messageSub();
  };
}

export async function stopPush(): Promise<void> {
  const factory = await loadMessaging();
  unsubscribe?.();
  unsubscribe = null;
  if (!factory || !token) return;
  const lastToken = token;
  token = null;
  try {
    await apiClient.unregisterPushToken(lastToken);
    await factory().deleteToken();
  } catch (error) {
    telemetry.report(error instanceof Error ? error : new Error('push unregister failed'));
  }
}

File: flutter/lib/push/push_service.dart
// 推送基线：FCM/APNs 令牌注册 + 前台消息转发。
// 依赖原生 Firebase 配置（google-services / GoogleService-Info），与 telemetry
// 同策略：未配置平台静默降级（令牌获取/权限失败不抛出、不阻塞登录流程）。
import 'dart:async';

import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_messaging/firebase_messaging.dart';

import '../telemetry/telemetry.dart';

/// 后台消息回调要求顶层函数；模板基线仅保活入口（业务可在此补埋点/本地库写入）。
@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {}

class PushService {
  static String? _token;
  static bool _starting = false;
  static StreamSubscription<RemoteMessage>? _messageSub;
  static StreamSubscription<String>? _tokenSub;
  static Future<bool> Function(String token)? _register;
  static Future<void> Function(String token)? _unregister;
  static void Function(String message)? _onForeground;

  /// 登录后调用：请求权限、取令牌上报服务端，并监听令牌轮换与前台消息。
  static Future<void> start({
    required Future<bool> Function(String token) register,
    required Future<void> Function(String token) unregister,
    required void Function(String message) onForeground,
  }) async {
    if (_starting || _messageSub != null) return;
    _starting = true;
    _register = register;
    _unregister = unregister;
    _onForeground = onForeground;
    try {
      if (Firebase.apps.isEmpty) await Firebase.initializeApp();
      final messaging = FirebaseMessaging.instance;
      await messaging.requestPermission();
      await _registerToken(await messaging.getToken());
      _tokenSub = messaging.onTokenRefresh.listen(_registerToken);
      _messageSub = FirebaseMessaging.onMessage.listen((message) {
        final body = message.notification?.body;
        if (body != null && body.isNotEmpty) _onForeground?.call(body);
      });
      FirebaseMessaging.onBackgroundMessage(firebaseMessagingBackgroundHandler);
    } catch (error, stack) {
      // 原生配置缺失/模拟器无 APNs 等：推送整体降级，telemetry 留痕即可。
      telemetry.report(error, stack);
    } finally {
      _starting = false;
    }
  }

  static Future<void> _registerToken(String? token) async {
    if (token == null || token.isEmpty) return;
    _token = token;
    try {
      await _register?.call(token);
    } catch (error, stack) {
      telemetry.report(error, stack);
    }
  }

  /// 登出调用：解绑监听、服务端解除令牌绑定并删除本地令牌。
  static Future<void> stop() async {
    await _messageSub?.cancel();
    _messageSub = null;
    await _tokenSub?.cancel();
    _tokenSub = null;
    final token = _token;
    _token = null;
    if (token == null) return;
    try {
      await _unregister?.call(token);
      if (Firebase.apps.isNotEmpty) {
        await FirebaseMessaging.instance.deleteToken();
      }
    } catch (error, stack) {
      telemetry.report(error, stack);
    }
  }
}
